File: Python/279__Perfect_Squares.py
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        squares = [i**2 for i in range(1, int(math.sqrt(n) + 1))]
        thisVisit = set()
        thisVisit.add(0)
        toVisit = set()
        bfsRound = 0

        while True:
            bfsRound += 1
            for num in thisVisit:
                for square in squares:
                    if num + square < n:
                        toVisit.add(num + square)
                    elif num + square == n:
                        return bfsRound
                    else:
                        break
            thisVisit, toVisit = toVisit, set()
        

File: Java/279__Perfect_Squares.java
class Solution {
    public int numSquares(int n) {
        List<Integer> perfSquares = new ArrayList<>();
        for (int i = 1; i * i <= n; i++)
            perfSquares.add(i * i);
        int[] dp = new int[n + 1];
        dp[0] = 0;
        for (int i = 1; i <= n; i++) {
            for (int num : perfSquares) {
                if (i - num < 0)
                    break;
                else
                    if (dp[i] == 0 || dp[i] > dp[i - num] + 1)
                        dp[i] = dp[i - num] + 1;
            }
        }
        return dp[n];
    }
}
